File: Demos/OC/UFileSDKDemo/UFileSDKDemo/file/UFMultiPartCell.m
//
//  UFMultiPartCell.m
//  UFileAssistant
//
//  Created by ethan on 2018/11/15.
//  Copyright © 2018 ucloud. All rights reserved.
//

#import "UFMultiPartCell.h"

@interface UFMultiPartCell()
@property (assign, nonatomic) BOOL   bUploaded;
@property (weak, nonatomic) IBOutlet UIProgressView *progress;
@property (weak, nonatomic) IBOutlet UIButton *btnUpload;
@property (nonatomic, strong) UFDataManager* dataManager;

@end

@implementation UFMultiPartCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    self.progress.progress = .0;
    self.progress.progressTintColor = [UIColor blueColor];
    
}

-(void)setDataManager:(UFDataManager *)dataManager ufFileClient:(UFFileClient *)flientClient PartNumber:(NSInteger)partnumber
{
    self.dataManager = dataManager;
    self.ufClient = flientClient;
    self.partNumber = partnumber;
    
    NSString *etag = [self.dataManager.etags objectForKey:[NSString stringWithFormat:@"%ld",partnumber]];
    if (!etag) {
        self.progress.progress = .0;
        self.btnUpload.enabled = YES;
    }else{
        self.progress.progress = 1;
        self.btnUpload.enabled = NO;
    }
//    if (!self.bUploaded) {
//        self.progress.progress = .0;
//        self.btnUpload.enabled = YES;
//    }
}

- (IBAction)onpressedButtonUpload:(id)sender {
    __weak typeof(self) weakself = self;
    
//    NSLog(@"%@,---data leng:%d",self.dataManager.multiPartInfo,(int)([self.dataManager getDataForPart:self.partNumber].length) );
    
    /*
     // 支持后台上传
     NSData *data = [self.dataManager getDataForPart:self.partNumber];
     NSString *filePath = [self.dataManager writeData:data fileName:@"video"];
     
     [self.ufClient startMultipartUploadWithKeyName:self.dataManager.multiPartInfo.key mimeType:@"video/quicktime" uploadId:self.dataManager.multiPartInfo.uploadId partIndex:self.partNumber dataLength: [data length] filePath:filePath progress:^(NSProgress * _Nonnull progress) {
         dispatch_async(dispatch_get_main_queue(), ^{
             [weakself.progress setProgress:progress.fractionCompleted animated:YES];
         });
     } uploadHandler:^(UFError * _Nullable ufError, UFUploadResponse * _Nullable ufUploadResponse) {
         if (!ufError) {
             weakself.bUploaded = YES;
             weakself.btnUpload.enabled = NO;
             [self.dataManager addEtag:ufUploadResponse.etag partNumber:ufUploadResponse.partNumber];
            // 上传成功，需要清理本地存储的临时文件
             return;
         }
         dispatch_async(dispatch_get_main_queue(), ^{
             weakself.progress.progressTintColor = [UIColor redColor];
         });
     }];
     */
    
    [self.ufClient startMultipartUploadWithKeyName:self.dataManager.multiPartInfo.key mimeType:@"video/quicktime" uploadId:self.dataManager.multiPartInfo.uploadId partIndex:self.partNumber fileData:[self.dataManager getDataForPart:self.partNumber] progress:^(NSProgress * _Nonnull progress) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakself.progress setProgress:progress.fractionCompleted animated:YES];
        });
    } uploadHandler:^(UFError * _Nullable ufError, UFUploadResponse * _Nullable ufUploadResponse) {
        if (!ufError) {
            weakself.bUploaded = YES;
            weakself.btnUpload.enabled = NO;
            [self.dataManager addEtag:ufUploadResponse.etag partNumber:ufUploadResponse.partNumber];
            
            return;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            weakself.progress.progressTintColor = [UIColor redColor];
        });
    }];
}


- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

@end

File: Demos/Swift/UFileSDKDemo-swift/UFileSDKDemo-swift/file/UFMultiPartCell.swift
//
//  UFMultiPartCell.swift
//  UFileSDKDemo-swift
//
//  Created by ethan on 2018/12/13.
//  Copyright © 2018 ucloud. All rights reserved.
//

import UIKit

class UFMultiPartCell: UITableViewCell {

    @IBOutlet weak var progress:UIProgressView!
    @IBOutlet weak var btnUpload:UIButton!
    
    var partNumber:Int = 0
    var ufFileClient:UFFileClient = UFFileClient()
    var dataManager:UFDataManager? = nil
    var bUploadded:Bool = false
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.progress.progress = 0
        self.progress.progressTintColor = UIColor.blue
    }
    
    public func setDataManager(dataManager:UFDataManager, ufFileClient:UFFileClient, partNumber:Int) -> Void {
        self.dataManager = dataManager
        self.ufFileClient = ufFileClient
        self.partNumber = partNumber
        
        let etag = self.dataManager!.etags.object(forKey: String(partNumber))
        if etag == nil {
            self.progress.progress = 0
            self.btnUpload.isEnabled = true
            return
        }
        self.progress.progress = 1
        self.btnUpload.isEnabled = false
    }
    
    @IBAction func onpressedButtonUpload(_ sender: Any)
    {
        weak var weakself:UFMultiPartCell! = self
        self.ufFileClient.startMultipartUpload(withKeyName: (self.dataManager!.multiPartInfo?.key)!, mimeType: "video/quicktime", uploadId: self.dataManager!.multiPartInfo!.uploadId, part: self.partNumber, fileData: self.dataManager!.getDataForPart(partNumber: UInt(self.partNumber)), progress: { (progress:Progress) in
            
            DispatchQueue.main.async {
                weakself.progress.setProgress(Float(progress.fractionCompleted), animated: true)
            }
            
        }) { (ufError:UFError?, ufUploadResponse:UFUploadResponse?) in
            if ufError == nil {
                weakself.bUploadded = true
                weakself.btnUpload.isEnabled = false
                weakself.dataManager!.addEtag(etag: (ufUploadResponse?.etag)!, partNumber: Int(ufUploadResponse!.partNumber))
                return
            }
            
            DispatchQueue.main.async {
                weakself.progress.progressTintColor = UIColor.red
            }
        }
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

}
